src/Vector3d.h

```diff
@@ -0,0 +1,218 @@
+            if (_v[1] == v._v[1]) {
+                return _v[2] < v._v[2];
+            }
+            return _v[1] < v._v[1];
+        }
+        return _v[0] < v._v[0];
+    }
+
+    ///@{
+    /// `data` returns a pointer to the 3 components of this vector.
+    double * data() { return _v; }
+    double const * data() const { return _v; }
+    ///@}
+
+    ///@{
+    /// The function call operator returns the `i`-th component of this vector.
+    double & operator()(int i) { return _v[i]; }
+    double operator()(int i) const { return _v[i]; }
+    ///@}
+
+    ///@{
+    /// `x` returns the first component of this vector.
+    double & x() { return _v[0]; }
+    double x() const { return _v[0]; }
+    ///@}
+
+    ///@{
+    /// `y` returns the second component of this vector.
+    double & y() { return _v[1]; }
+    double y() const { return _v[1]; }
+    ///@}
+
+    ///@{
+    /// `z` returns the third component of this vector.
+    double & z() { return _v[2]; }
+    double z() const { return _v[2]; }
+    ///@}
+
+    /// `dot` returns the inner product of this vector and `v`.
+    double dot(Vector3d const & v) const {
+        return _v[0] * v._v[0] + _v[1] * v._v[1] + _v[2] * v._v[2];
+    }
+    /// `squaredNorm` returns the inner product of this vector with itself.
+    double squaredNorm() const { return dot(*this); }
+    /// `norm` returns the L2 norm of this vector.
+    double norm() const {
+        return std::sqrt(squaredNorm());
+    }
+
+    /// `normalize` scales this vector to have unit norm and returns its norm
+    /// prior to scaling. It will accurately normalize any vector with finite
+    /// components except for (0, 0, 0), including those with norms that
+    /// overflow. Trying to normalize (0, 0, 0) will cause a std::runtime_error
+    /// to be thrown.
+    double normalize() {
+        double scale = 1.0;
+        double invScale = 1.0;
+        double n2 = squaredNorm();
+        if (n2 < 4.008336720017946e-292) {
+            // If n2 is below 2^(-1022 + 54), i.e. close to the smallest normal
+            // double precision value, scale each component by 2^563 and
+            // recompute the squared norm.
+            scale = 3.019169939857233e+169;
+            invScale = 3.312168642111238e-170;
+            n2 = ((*this) * scale).squaredNorm();
+            if (n2 == 0.0) {
+                throw std::runtime_error("Cannot normalize zero vector");
+            }
+        } else if (n2 == std::numeric_limits<double>::infinity()) {
+            // In case of overflow, scale each component by 2^-513 and
+            // recompute the squared norm.
+            scale = 3.7291703656001034e-155;
+            invScale = 2.6815615859885194e+154;
+            n2 = ((*this) * scale).squaredNorm();
+        }
+        double norm = std::sqrt(n2);
+        _v[0] = (_v[0] * scale) / norm;
+        _v[1] = (_v[1] * scale) / norm;
+        _v[2] = (_v[2] * scale) / norm;
+        return norm * invScale;
+    }
```

Comment:
Should go in .cc file.

I also wonder how much you really want to "properly" handle normalization of truly tiny vectors; I would guess that most of the time a vector as tiny as the ones you're worried about here would have actually been zero in exact arithmetic, and throwing an exception is just as appropriate - if not ore appropriate - than dealing with denormalized floating point values.